### PR TITLE
Revert if not enough gas

### PR DESCRIPTION
**Pull request by anxolin**
Makes trampoline contract to revert if there's no enough gas left

## Context
Solvers might under-estimate the gas required for a settlement with hooks.

Hooks have a `gasLimit` that they require, however, the settlement transaction can indicate a low gas, so when we arrive to the hook the gas left is less than the `gasLimit`. 
This will likely result in the hook not getting some fair gas to execute and failing. 

This goes unnoticed for the solver, because the settlement will succeed. 

This PR makes sure the gas left in the transaction is enough to honor the instructed `gasLimit`, if not, it reverts. This would make it so the solvers get a revert in the settlement simulation if this is the case.

## Changes
Before each hook execution, we check there's enough gas left:

`require(gasleft() >= hook.gasLimit, "Not enough gas");`

## Unit tests
Added 2 tests:
- `test_RevertsWhenNotEnoughGas`
- `test_RevertsWhenNotEnoughGasForMultipleHooks`

## Test
```
forge test
```

### CODE REVIEW

**Review by carlos-cow — DISMISSED** [review]

File: test/HooksTrampoline.t.sol [hunk 1]
```diff
@@ -160,3 +198,11 @@ contract Hummer {
         }
     }
 }
+
+contract BurnGas {
+    function consumeAllGas() public {
+        while (true) {
```
Comment (carlos-cow) [inline test/HooksTrampoline.t.sol:204]: does this finish fast enough?

You can also try to replace with this https://github.com/Vectorized/solady/blob/main/src/utils/GasBurnerLib.sol#L49
for faster gas burn

Comment (carlos-cow) [reply test/HooksTrampoline.t.sol:204]: apparently it runs fast enough, nevermind
```
Ran 1 test suite in 6.59ms (1.19ms CPU time): 8 tests passed, 0 failed, 0 skipped (8 total tests)
```

File: test/HooksTrampoline.t.sol [hunk 2]
```diff
@@ -160,3 +198,11 @@ contract Hummer {
...
+            // Do nothing, just consume gas
+        }
+    }
+}
```
[the remark this answers is not in the document]
Comment (anxolin) [reply test/HooksTrampoline.t.sol]: 9b8e2b02f8e30fa34e773efd26fc56a1f4772440

File: src/HooksTrampoline.sol [hunk 3]
```diff
@@ -55,6 +55,7 @@ contract HooksTrampoline {
             Hook calldata hook;
             for (uint256 i; i < hooks.length; ++i) {
                 hook = hooks[i];
+                require(gasleft() >= hook.gasLimit, "Not enough gas");
```
[the remark this answers is not in the document]
Comment (anxolin) [reply src/HooksTrampoline.sol]: I've made the test to be isolate=true (except one that was breaking).

Not sure if this is what you would want, but we could do something to 
```solidity
uint256 effectiveGas = gasleft() * 63 / 64;
if (effectiveGas < hook.gasLimit) {
    revert NotEnoughGas();
}
```

Yes, I guess your concern is valid. Someone that want to not break the rules, could technically craft a transaction that leaves just too little gas, this 1/64 eams my tests above passes, but the hook does't get enough gas (compared to the limit).

However, is this really so bad and worth it to overcomplicate the logic? I would think, as long as hooks gas are over-estimated (by 1.57%) then its not possible to cause this. 

Assuming I'm getting it, I would be inclined to go with this version tomorrow live

**Comment by anxolin** [comment]
@fedgiac @carlos-cow  I addressed most of the comments

Also, I added all the tests you suggested in the chat under `GasLimitEnforcementTest` (I just broke them down and refactored them). 

I also tested that indeed 1/64 of the remainder is not forwarded. See `test_DemonstratesOneSixtyFourthGasReservation` . I did not create a test that reproduces the edge case, if important I can try to reproduce tomorrow. I would also have some time tomorrow to review more failed transactions before swap and bridge goes live.

**Review by fedgiac — COMMENTED** [review]
My comments are in https://github.com/cowprotocol/hooks-trampoline/pull/15.

File: src/HooksTrampoline.sol [hunk 4]
```diff
@@ -55,6 +63,12 @@ contract HooksTrampoline {
             Hook calldata hook;
             for (uint256 i; i < hooks.length; ++i) {
                 hook = hooks[i];
+                // A call forwards all but 1/64th of the available gas. The
+                // math is used as a heuristic to account for this.
+                uint256 forwardedGas = gasleft() * 63 / 64;
+                if (forwardedGas < hook.gasLimit) {
+                    revert NotEnoughGas();
```
Comment (fleupold) [inline src/HooksTrampoline.sol:70]: I'm afraid this implementation suffers from breaking the default `eth_estimateGas` implementation of many RPC nodes. Given this is a vanilla revert (not a real running out of gas) RPCs don't know that they can avoid this by increasing the gas estimate.

One way to repro is to try to estimate gas for a hook that uses a gasLimit of 10M (it will just say that it reverts). Do you think you can come up with an implementation where a hook with this gas limit can pass the `estimateGas` call for most nodes (reth, geth, erigon)?

**Comment by anxolin** [comment]
@fedgiac can I mege? I want to mege this to main, and then https://github.com/cowprotocol/hooks-trampoline/pull/18

**Review by fedgiac — DISMISSED** [review]
Looks good! (Comments are just nits, feel free to ignore.)
